src/client/datascience/jupyter/jupyterExecution.ts

```
@@ -153,11 +155,38 @@ export class JupyterExecutionBase implements IJupyterExecution {
                         enableDebugging: options ? options.enableDebugging : false
                     };
 
-                    traceInfo(`Connecting to process for ${options ? options.purpose : 'unknown type of'} server`);
-                    await result.connect(launchInfo, cancelToken);
-                    traceInfo(`Connection complete for ${options ? options.purpose : 'unknown type of'} server`);
+                    // tslint:disable-next-line: no-constant-condition
+                    while (true) {
+                        try {
+                            traceInfo(`Connecting to process for ${options ? options.purpose : 'unknown type of'} server`);
+                            await result.connect(launchInfo, cancelToken);
+                            traceInfo(`Connection complete for ${options ? options.purpose : 'unknown type of'} server`);
+                            break;
+                        } catch (ex) {
+                            if (ex instanceof JupyterSessionStartError && isLocalConnection) {

```

>instanceof [](start = 35, length = 10)

Why not let this bubble out to start server in the interactiveBase?

---

Because we wouldn't want to start a server all over again. Besides we are already performing a selection of kernels in this same method (earlier on).

---

Okay. I'm just anxious about UI proliferating into server code. Now a server component has a dependency on talking to the user.

---
In reply to: [360156416](https://github.com/microsoft/vscode-python/pull/9227#discussion_r360156416) [](ancestors = 360156416)

---

I agree. But I think we should probably break this into a separate class. Like `NotebookStarter` or similar so everything to do with starting notebooks/sessions is in there.